src/actions/GoalHandle.js

```diff
@@ -0,0 +1,182 @@
+/*
+ *    Copyright 2017 Rethink Robotics
+ *
+ *    Copyright 2017 Chris Smith
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+'use strict';
+
+const msgUtils = require('../utils/message_utils.js');
+const timeUtils = require('../lib/Time.js');
+
+let GoalStatus = null;
+let GoalStatuses = null;
+
+class GoalHandle {
+  constructor(goalId, actionServer, status) {
+    if (goalId.id === '') {
+      goalId = actionServer.generateGoalId();
+    }
+
+    if (timeUtils.isZeroTime(goalId.stamp)) {
+      goalId.stamp = timeUtils.now();
+    }
+
+    this.id = goalId.id;
+
+    this._as = actionServer;
+
+    if (GoalStatus === null) {
+      GoalStatus = msgUtils.requireMsgPackage('actionlib_msgs').msg.GoalStatus;
+      GoalStatuses = GoalStatus.Constants;
+    }
+
+    this._status = new GoalStatus({
+      status: status || GoalStatuses.PENDING,
+      goal_id: goalId
+    });
+
+    this._destructionTime = timeUtils.epoch();
+  }
+
+  _isTerminalState() {
+    return [
+      GoalStatuses.REJECTED,
+      GoalStatuses.RECALLED,
+      GoalStatuses.PREEMPTED,
+      GoalStatuses.ABORTED,
+      GoalStatuses.SUCCEEDED
+    ].includes(this._status.status);
+  }
+
+  getStatusId() {
+    return this._status.status;
+  }
+
+  getGoalId() {
+    return this._status.goal_id;
+  }
+
+  getGoalStatus() {
+    return this._status;
+  }
+
+  publishFeedback(feedback) {
+    this._as.publishFeedback(this._status, feedback);
+  }
+
+  _setStatus(status, text) {
+    this._status.status = status;
+    if (text) {
+      this._status.text = text;
+    }
+
+    // FIXME: just guessing about setting destruction time
+    if (this._isTerminalState()) {
+      this._destructionTime = timeUtils.now();
+    }
+
+    this._as.publishStatus();
+  }
+
+  _publishResult(result) {
+    this._as.publishResult(this._status, result);
+  }
+
+  setCancelled(result, text = '') {
+    const status = this.getStatusId();
+    switch (status) {
+      case GoalStatuses.RECALLING:
+        this._setStatus(GoalStatuses.RECALLED, text);
+        this._publishResult(result);
+        break;
+      case GoalStatuses.PREEMPTING:
+        this._setStatus(GoalStatuses.PREEMPTED, text);
+        this._publishResult(result);
+        break;
+      default:
+        // FIXME: log errors for invalid state transitions
```

Comment:
As a reference, perhaps toss in this link:
http://wiki.ros.org/actionlib/DetailedDescription#Server_Description
to aid the future fixer :)

Reply:
In the mean time, just logging any error about an unknown transition would be helpful...